### DISCUSSION TITLE
How to create refs in App router page?

### DISCUSSION THREAD

**Discussion by sushantgwr87** [vercel/next.js#60042] [Help] [answered]
### Summary

Hello,

I have been using page router of NextJS most of the time but this time wanted to test-out App Routing. Most of thing I understood from tutorials, issue answers and discussions.

## Problem
I am creating a page which has animations on elements, those animations are triggered using intersection observer hook which takes refs as parameter and observe it, then return true or false accordingly.

Problem is I can't use useRef on server side and for that I need to make the page itself client-side saying "use client" at top.

So questions are:
1) Is there any other way to add refs in server component? I am talking here to add multiple refs to different elements of same Page. Or any other way around to animate such elements.
2) Using "use client" on page will be a good option? Like I clearly can't grasp this situation. I think using "use client" just removes the reason behind using app router and won't be that good but then it make things easy as well.

I am trying to learn this App Router currently so not well versed with situations and conditions I should consider before using this mix blend of client and server components.

### Additional information

```js
This is the useIntersectionObserver Hook I am using, it take array of refs:

import { useEffect, useRef, useState } from 'react';

const useIntersectionObserver = (elementRefs, rootMargin = '0px', threshold = 0) => {
  const [isIntersecting, setIsIntersecting] = useState(() => Array(elementRefs.length).fill(false));
  const observerRefs = useRef([]);

  useEffect(() => {
    const options = {
      root: null,
      rootMargin,
      threshold,
    };

    const callbacks = elementRefs.map((_, index) => (entries) => {
      entries.forEach((entry) => {
        if (entry.target === elementRefs[index].current) {
          setIsIntersecting((prev) => {
            const updatedIntersecting = [...prev];
            updatedIntersecting[index] = entry.isIntersecting;
            return updatedIntersecting;
          });
        }
      });
    });

    // Create an Intersection Observer instance for each element
    const observers = elementRefs.map((elementRef, index) => {
      const observer = new IntersectionObserver(callbacks[index], options);
      observer.observe(elementRef.current);
      return observer;
    });

    // Store the observer instances in the ref for cleanup
    observerRefs.current = observers;

    // Cleanup the observers on unmount
    return () => {
      observerRefs.current.forEach((observer) => observer.disconnect());
    };
  }, [elementRefs, rootMargin, threshold]);

  return { isIntersecting };
};

export default useIntersectionObserver;
```

**Comment by ananuness** [comment, marked as answer]
Hi, Sushant!! Here we go:

#### 1. Is there any other way to add refs in server component? I am talking here to add multiple refs to different elements of same Page. Or any other way around to animate such elements.
In server components you can't use any code that requires browser environment. With that said, you can't use any of those hooks directly there, except for some "hooks" that are server side.

#### 2. Using "use client" on page will be a good option? Like I clearly can't grasp this situation. I think using "use client" just removes the reason behind using app router and won't be that good but then it make things easy as well.
Definitely is an easier resolution, but as you said, it removes the reason to use SSR. But a good thing to know is that you can render Client components in you Server component/page!! So, one good thing you can do to take advantage of SSR and cia, is separate the elements that need client rendering into Client components and import them on your server-side page! 

Therefore, your `useIntersectionObserver` hook will be client side, but you can call it within a Server component/page. Just be careful to separate the elements that require client rendering into client components and you will make good use of both sides.

I hope it helps, any question don't hesitate to ask! 😀

**Reply by sushantgwr87** [reply]
Thanks for answer and explanation.

Got it, separating only those elements which I want animate to client side and then importing them seems better than declaring whole page as client side. Though it will be a good headache to do that 😅 as was used to do in 1 page in page routing but need to do it.

Again, thanks for answer. 🥂

**Reply by ananuness** [reply]
You're welcome!! 🥂
